packages/governance/src/governor/extensions/governor_timelock_execution.cairo

```diff
@@ -0,0 +1,295 @@
+    type ProposalId = felt252;
+    type TimelockProposalId = felt252;
+
+    #[storage]
+    pub struct Storage {
+        pub Governor_timelock_controller: ContractAddress,
+        pub Governor_timelock_ids: Map<ProposalId, TimelockProposalId>
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    pub enum Event {
+        TimelockUpdated: TimelockUpdated,
+    }
+
+    /// Emitted when the timelock controller used for proposal execution is modified.
+    #[derive(Drop, starknet::Event)]
+    pub struct TimelockUpdated {
+        pub old_timelock: ContractAddress,
+        pub new_timelock: ContractAddress
+    }
+
+    mod Errors {
+        pub const INVALID_TIMELOCK_CONTROLLER: felt252 = 'Invalid timelock controller';
+    }
+
+    //
+    // Extensions
+    //
+
+    pub impl GovernorExecution<
+        TContractState,
+        +GovernorComponent::HasComponent<TContractState>,
+        +GovernorComponent::GovernorSettingsTrait<TContractState>,
+        +GovernorComponent::GovernorCountingTrait<TContractState>,
+        +GovernorComponent::GovernorVotesTrait<TContractState>,
+        +SRC5Component::HasComponent<TContractState>,
+        impl GovernorCoreExecution: HasComponent<TContractState>,
+        +Drop<TContractState>
+    > of GovernorComponent::GovernorExecutionTrait<TContractState> {
+        /// See `GovernorComponent::GovernorExecutionTrait::state`.
+        fn state(
+            self: @GovernorComponentState<TContractState>, proposal_id: felt252
+        ) -> ProposalState {
+            let current_state = self._state(proposal_id);
+
+            if current_state != ProposalState::Queued {
+                return current_state;
+            }
+
+            let contract = self.get_contract();
+            let this_component = GovernorCoreExecution::get_component(contract);
+
+            let queue_id = this_component.Governor_timelock_ids.read(proposal_id);
+            let timelock_controller = this_component.timelock();
+            let timelock_dispatcher = ITimelockDispatcher { contract_address: timelock_controller };
+
+            if timelock_dispatcher.is_operation_pending(queue_id) {
+                ProposalState::Queued
+            } else if timelock_dispatcher.is_operation_done(queue_id) {
+                // This can happen if the proposal is executed directly on the timelock.
+                ProposalState::Executed
+            } else {
+                // This can happen if the proposal is canceled directly on the timelock.
+                ProposalState::Canceled
+            }
+        }
+
+        /// See `GovernorComponent::GovernorExecutionTrait::executor`.
+        ///
+        /// In this module, the executor is the timelock controller.
+        fn executor(self: @GovernorComponentState<TContractState>) -> ContractAddress {
+            let contract = self.get_contract();
+            let this_component = GovernorCoreExecution::get_component(contract);
+
+            this_component.timelock()
```

Comment:
Note that the self is not the current component state, but the GovernorComponent::ComponentState, since it is a extension. To access the current component storage we still need to access the current component (this_component)